Configurable probe termination grace periods

ehashman opened the pull request:
See #2238. Straw proposal to be discussed at the next SIG Node meeting.

/sig node
/cc @smarterclayton @derekwaynecarr

SergeyKanzhelev commented:
/cc: @matthyx

matthyx commented:
Thanks for the cc @SergeyKanzhelev !
/lgtm
Maybe Tim would like to comment on the api change too?
/cc @thockin

dims commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,491 @@
+set, and a liveness probe fails, a workload will not be promptly restarted
+because it will wait for the full termination period.
+
+I propose adding a new field to liveness probes, `gracePeriodSeconds`. When
```
Do we need to say that this field is not applicable to startupProbe and readinessProbe explicitly?

thockin replied:
I detest the existence of fields in re-used structs that can only be used in some contexts.  If we do this, KEP, can we please create a new struct (stasring with a copy of the old) for those places that need the new fields?

Also - wouldn't startupProbes use this?

ehashman replied:
So I sort of left this intentionally vague as a discussion point :)

I think that readiness probes use a different code path, but could be configured with this value if we wanted it to. I don't know if the bug currently affects startup probes, since they're pretty new.

matthyx replied:
I don't think this field is also needed for the `startupProbe` as it only terminates the container when the probe has never succeeded. So I don't think there is anything to cleanup which would require a long termination period.
Maybe I am wrong for some use cases where the `startupProbe` is used to cover a set of DB scripts?

ehashman replied:
I think `startupProbes` are affected. Upon probe failure, both liveness and startup probes queue the container to be killed: https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L631-L655 This is presumably why the `terminationGracePeriodSeconds` from the pod spec is used for termination.

Readiness probes go through a different code path (indicated by [this comment](https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/prober/prober_manager.go#L55)) although I haven't figured out the exact difference yet.

ehashman replied:
I've updated the KEP to explicitly call out liveness/startup probes, and not readiness.

ehashman replied:
(pasting this for reference here, this is where the grace termination periods are currently set upon container kill: https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/kuberuntime/kuberuntime_container.go#L604-L611)

dims commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,491 @@
+This change would be [API-compatible][compatible]. If the field is not
+specified, we will default to the current behaviour, which uses the
+`terminationGracePeriodSeconds`.
+
```
Can we add some discussion here on how this value relates to initialDelaySeconds/timeoutSeconds/periodSeconds and the failureThreshold?

Also does `gracePeriodSeconds` convey the meaning of the field? (since we say "configurable timeout on failure" in the goal above)

ehashman replied:
I am open to better names for the field.

dims commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,491 @@
+E2E integration tests will verify the correct behaviour for the current broken
+use case, where a `terminationGracePeriodSeconds` is set to a
+large value, a probe fails, and with `livenessProbe.gracePeriodSeconds` set,
+the container terminates quickly.
```
do we allow livenessProbe.gracePeriodSeconds to be higher than terminationGracePeriodSeconds? do we need some kinds of validation?

dims commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,491 @@
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
```
Do we need a feature gate?

ehashman replied:
I've not proposed anything like this before, so let me know if I'm off base here, but I think as a bugfix this is the sort of change that could graduate directly to stable. It's a compatible API change, so the code path won't be encountered unless the new field is used, and adding a feature gate around that seems like it is more likely to introduce bugs than prevent them.

ehashman commented:
/retest

logicalhan commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,491 @@
+
+E2E integration tests will verify the correct behaviour for the current broken
+use case, where a `terminationGracePeriodSeconds` is set to a
+large value, a probe fails, and with `livenessProbe.gracePeriodSeconds` set,
```
How about `livenessProbe.terminationGracePeriodSeconds`? The semantics of overriding configs with greater specificity should be pretty intuitive to most people I suspect.

dims commented:
/hold (plenty of feedback needs to be rolled in)

ehashman commented:
I think I've addressed all comments, please take another look @matthyx @dims @thockin @logicalhan

I will remove the hold since the LGTM has been removed :)

/hold cancel

derekwaynecarr commented:
will take review for sig-node.
/assign

smarterclayton commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+be terminated, update the behaviour of liveness probes to terminate the pod
+immediately, or with some default grace period (@sjenning)
+
+- PRO: avoids API change
```
I will add another CON: Kubernetes changes behavior in a surprising way, potentially breaking existing apps that depend on grace period for crash safety.  That CON is sufficient for me to NACK such a change, because the cost of breaking existing users in these scenarios is potentially very high and would happen in the worst scenario (terminationGracePeriod is intended to allow applications a bounded time window to shut down).

I just want to record that here so that in the future we treat crash safety under SIGTERM as it was intended all the way back in https://github.com/kubernetes/kubernetes/issues/2789 (https://github.com/kubernetes/kubernetes/pull/8641 is where we actually say "don't SIGKILL my database" in an issue, it was always implicit).

derekwaynecarr commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+terminating the container when a probe fails, and override the pod-level
+`terminationGracePeriodSeconds` value.
+
+This change would apply to [liveness and startup probes][probes1], but not
```
i know the desire is to only introduce this attribute on these two types of probes, but it has a cost to introduce a new probe type specific to readiness in kubelet as we treated all of these generically all over the place.

i actually prefer we ensure in validation that `terminationGracePeriodSeconds` remains unset in validation logic for readiness probes than introduce a new type in types.go as i really think a common type simplifies implementation in kubelet.  this is no different in my view of us requiring a specific success threshold for liveness probe.

/cc @thockin

derekwaynecarr commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+This change will be unit tested for the API changes and
+backwards-compatibility.
+
+If added to the Probe API, this will also need to be tested in the readiness
```
i prefer its added to Probe API versus a new type.

derekwaynecarr commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+
+### Graduation Criteria
+
+Because this is a bugfix with a compatible API change, similar to
```
I think this requires a feature gate, and should not be treated different than any other change.

Its a new feature (decouple timeout), whereas the referenced example was a bug in existing feature.

The new field requires careful rollout because of n-2 kubelet skew requirements.

The API can only be supported when we know all kubelets have the logic, so it needs to be off-by default for at least 2 releases until all kubelets are up to same level as API server.

smarterclayton replied:
Yeah, needs a feature gate, it's an API change.

derekwaynecarr commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+-->
+
+### Version Skew Strategy
+
```
version skew requirements for n-2 delta between API server that would accept the field and kubelet ability to support it requires this section get populated.  it also requires a feature gate as a consequence.

derekwaynecarr reviewed (changes requested):
Thanks for this write-up!

I like the new field, prefer its added as optional to existing Probe API, validation only supports it for two probe types in question, and that we require a feature gate to control rollout of this support as we need to ensure API server and kubelet skew around adding new API fields.

smarterclayton commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+      failureThreshold: 1
+      periodSeconds: 60
+      # New field #
+      gracePeriodSeconds: 60
```
I would prefer naming be identical - terminationGracePeriodSeconds.  That clearly identifies the two are related.

ehashman replied:
Oops, I just missed updating the example when I updated the rest of the doc. Thanks for catching.

smarterclayton commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,544 @@
+## Design Details
+
+The `livenessProbe.terminationGracePeriodSeconds` (or `startupProbe.*`) shall
+not be greater than the pod-level `terminationGracePeriodSeconds`, and we will
```
I do not think this is required.  I see no need to couple the two fields, since there is no downside to a 30s graceful termination for regular and a 60s liveness termination (workload has fast shutdown, but failure detection can take longer).

ehashman replied:
@dims wdyt?

dims replied:
@ehashman works for me!

ehashman commented:
@derekwaynecarr @smarterclayton hopefully I've addressed all your feedback, PTAL

thanks for bearing with me on my first API change :)

ehashman commented on keps/sig-node/2238-liveness-probe-grace-period/kep.yaml:
```diff
@@ -0,0 +1,32 @@
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+status: provisional
```
Let me know if I should update this to implementable @derekwaynecarr or if we want a second PR to give folks another chance to TAL before KEP freeze

ehashman replied:
I bumped it to implementable given how close we're getting to KEP freeze.

ehashman commented on keps/sig-node/2238-liveness-probe-grace-period/kep.yaml:
```diff
@@ -0,0 +1,32 @@
+approvers:
+  - "@derekwaynecarr"
+prr-approvers:
+  - "@johnbelamaric"
```
cc @johnbelamaric

johnbelamaric commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,569 @@
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate (also fill in values in `kep.yaml`)
```
Please also add in `kep.yaml` here: https://github.com/kubernetes/enhancements/blob/93431db331b1e550546cc81e02e2ed0017e4a01b/keps/NNNN-kep-template/kep.yaml#L42

johnbelamaric commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,569 @@
+set, and a liveness probe fails, a workload will not be promptly restarted
+because it will wait for the full termination period.
+
+I propose adding a new field to liveness probes, `gracePeriodSeconds`. When
```
in the Proposal section you call this `terminationGracePeriodSeconds` not `gracePeriodSeconds`.

ehashman replied:
Oops, missed one more rename.

johnbelamaric commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,569 @@
+  with and without the feature, are necessary. At the very least, think about
+  conversion tests if API types are being modified.
+
+  Yes. We will add an e2e test to confirm the presence of the current bug.
```
This section is about enablement/disablement. For example, if you create the resource with the new field, then you disable the feature gate, the field should not be used even though it is still present and filled in. Alas our test framework doesn't allow this sort of test easily so it would be a manual test.

ehashman replied:
Ah, I see. Thanks.

SergeyKanzhelev commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,569 @@
+> open connections), but when the controller wedges it takes an hour to
+> resolve, which is the worst possible outcome (the process wedges immediately,
+> but kube has to wait the full hour to kill it, defeating the goal of
+> liveness). [comment](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756201502)
```
is readiness probe helpful here? Theoretically readiness probe may change the status of the pod so it will be taken out of LB? And in non-ready state it can drain the connections while another ready pod will be scheduled.

SergeyKanzhelev replied:
if this is not the case, maybe it can be explained "why" in text. This was always my understanding of a difference of readiness and liveness probes

ehashman replied:
cc @smarterclayton, this example was from his comment; I think that we specifically want a liveness and not a readiness probe for this use case.

derekwaynecarr replied:
@SergeyKanzhelev usage of a readiness probe is orthogonal since the primary issue is a failed liveness or startup probe should not require the normal graceful termination period.

SergeyKanzhelev commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,569 @@
+We can add a `terminationGracePeriodSeconds` integer field to the Probe API.
+This field would specify how long the kubelet should wait before forcibly
+terminating the container when a probe fails, and override the pod-level
+`terminationGracePeriodSeconds` value.
```
would it be confusing to have value `terminationGracePeriodSeconds` specified in the probe? Example given in the beginning where `terminationGracePeriodSeconds` is used to drain nodes works fine for this new field. What if the `terminationGracePeriodSeconds` is used to make sure that the state is successfully recorded in some database on exit and designed by app developer while liveness probe is designed by "operations". It may be more logical to see this new field on a top level next to the exiting field. Maybe list this into the alternative considerations with explanaition why the new field is under the probe, not on the top level.

ehashman commented:
/label tide/merge-method-squash

SergeyKanzhelev commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,577 @@
+`terminationGracePeriodSeconds` value.
+
+This change would apply to [liveness and startup probes][probes1], but **not**
+readiness probes, which [use a different code path][probes2].
```
this is being changed by @matthyx I believe as we speak

derekwaynecarr replied:
the readiness probe cannot trigger a stop of container....

SergeyKanzhelev commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,577 @@
+### Goals
+
+- Liveness probes can have a configurable timeout on failure.
+- Existing behaviour is not changed.
```
is there any chance to get a race between the termination on node shutdown and liveness probe failure. Let's say liveness probe is set up to be extremely sensitive to the termination signal that pod receives. So pod started termination because it received sig kill, liveness probe immediately failed, and instead of regular interval, pod assumes liveness probe's interval override. So depending whether liveness probe checked in the period between termination started and finished, grace period will be different.

ehashman replied:
The termination paths for pod deletion/node drain and liveness probe failures are sort of separate (they use the same underlying function calls to kill the containers, but pod deletion always includes a hard override for the termination grace period). I don't think this would introduce any greater chance of a race condition being encountered.

Is it possible right now for a liveness probe to trigger specific container termination during a pod already in process of terminating? That's the only scenario I can think of that might pose an issue.

ehashman replied:
@matthyx is fixing this exact scenario, so it will no longer be possible: https://github.com/kubernetes/kubernetes/pull/98571

derekwaynecarr replied:
agree we should not probe for startup/liveness when a pod is undergoing graceful termination.  fwiw, if you send multiple stop calls to a container today, both primary runtimes hold a lock that prevents the subsequent call from running or reducing the termination period.  this is being fixed as well https://github.com/kubernetes/kubernetes/pull/98507/files

SergeyKanzhelev commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,577 @@
+Introduce a similar field, but at the pod-level, next to the existing field.
+
+- PRO: similar values are grouped together
+- CON: logically, liveness probes operate on a per-container basis, and thus we
```
so the grace period will sum up grace periods of all containers? I didn't get this CON rationale =(

ehashman replied:
The pod-level termination grace period makes sense to me because that's the overall time it should take to shut down the whole pod as a group of containers. But liveness probes don't act at the pod level, they act exclusively at the container level. So it doesn't make as much sense to me to set a global liveness timeout: what if you wanted to set them to different values for different containers?

SergeyKanzhelev replied:
so this is the time to restart the specific container in pod? What if the policy is restart never and pod needs to be terminated. Which timeout will be used?

ehashman replied:
Yes, liveness probes are specific to each container in a pod.

If the policy is restart never and the whole pod needs to be terminated, the other containers would either use the default grace period or, if set, the pod-level terminationGracePeriodSeconds.

SergeyKanzhelev replied:
that explains it. I misunderstood the KEP =) Will read again with this understanding.

Immediate question though, would users expect that liveness grace periods will be respected for each pod when overall pod needs to be terminated?

ehashman replied:
Nope. The idea of this KEP is to differentiate between the grace period for a regular, graceful shutdown when pod termination is requested (happy path), and a liveness probe failure (sad path), because these have been accidentally coupled.

johnbelamaric commented on keps/sig-node/2238-liveness-probe-grace-period/README.md:
```diff
@@ -0,0 +1,577 @@
+
+  We will manually test that if a resource is created with the new field, and
+  the feature flag is later disabled, that the field will no longer be used even
+  though it is already filled in.
```
Please also confirm that when you then re-enable the feature, the new value is used as expected.

ehashman commented:
@derekwaynecarr can you PTAL for final approval?

derekwaynecarr reviewed (approved):
/approve
/lgtm

ehashman commented:
@johnbelamaric can you do the final approval now?

johnbelamaric commented:
/approve
